myPDF.cxx

```diff
@@ -1,241 +1,255 @@
-     return 0.;
-
-   if ((mKP < MKaon + MPion) || (mKP > MBd - MPsi_nS) || (mPsiP < MPsi_nS + MPion) || (mPsiP > MBd - MKaon))
-     return 0.;
-
-   // cos(theta(K*)) from masses
-   Double_t mKP2 = mKP*mKP;
-   Double_t mPsiP2 = mPsiP*mPsiP;
-   Double_t MPsi_nS2 = MPsi_nS*MPsi_nS;
-   Double_t _cKs = cosTheta_FromMasses_host(mKP2, mPsiP2, MPsi_nS2, MBd2, MKaon2, MPion2);
-   //cout <<"_cKs = " <<_cKs <<" for mKP2 = " <<mKP2 <<" and mPsiP2 = " <<mPsiP2 <<endl;
-   
-   if (fabs(_cKs) > 1) {
-     //cout <<"_cKs = " <<_cKs <<" for mKP2 = " <<mKP2 <<" and mPsiP2 = " <<mPsiP2 <<endl;
-     return 0.;
-   }
- 
-   Double_t _phiTrue = phi; 
-   if (B0beauty < 0)
-     _phiTrue *= -1;
-
-   return PDF(_cKs,_phiTrue);
- }
-
- Int_t myPDF::getAnalyticalIntegral(RooArgSet& allVars, RooArgSet& analVars, const char* ) const //rangeName
- { 
-   // LIST HERE OVER WHICH VARIABLES ANALYTICAL INTEGRATION IS SUPPORTED, 
-   // ASSIGN A NUMERIC CODE FOR EACH SUPPORTED (SET OF) PARAMETERS 
-   // THE EXAMPLE BELOW ASSIGNS CODE 1 TO INTEGRATION OVER VARIABLE X
-   // YOU CAN ALSO IMPLEMENT MORE THAN ONE ANALYTICAL INTEGRAL BY REPEATING THE matchArgs 
-   // EXPRESSION MULTIPLE TIMES
-
-    // if (matchArgs(allVars,analVars,mKP)) return 1 ;
-    // if (matchArgs(allVars,analVars,mKP,cJ)) return 2 ;
-
-     return 0 ;
- } 
+myPDF::myPDF(const myPDF& other, const char* name) :
+  RooAbsPdf(other,name),
+  // B^0 -> psi(nS) K* -> mu+ mu- K- pi+
+  mKP("mKP",this,other.mKP),
+  cJ("cJ",this,other.cJ),
+  mPsiP("mPsiP",this,other.mPsiP),
+  phi("phi",this,other.phi),
+  B0beauty("B0beauty",this,other.B0beauty),
+  //
+  MPsi_nS(other.MPsi_nS),
+  //cKs(other.cKs),
+  Kstar_spin(other.Kstar_spin),
+  Zc_spin(other.Zc_spin),
+  varNames(other.varNames),
+  amplitudeVars(other.amplitudeVars),
+  amplitudeVarProxy_map(other.amplitudeVarProxy_map),
+  psi_nS(other.psi_nS),
+  dRadB0(other.dRadB0), dRadKs(other.dRadKs)
+{
+}
 
 
- Double_t myPDF::analyticalIntegral(Int_t code, const char* rangeName) const  
- { 
-   // RETURN ANALYTICAL INTEGRAL DEFINED BY RETURN CODE ASSIGNED BY getAnalyticalIntegral
-   // THE MEMBER FUNCTION x.min(rangeName) AND x.max(rangeName) WILL RETURN THE INTEGRATION
-   // BOUNDARIES FOR EACH OBSERVABLE x
+Double_t myPDF::evaluate() const
+{
+  // ENTER EXPRESSION IN TERMS OF VARIABLE ARGUMENTS HERE
+  if (fabs(cJ) > 1  ||  fabs(phi) > TMath::Pi())
+    return 0.;
+    
+  if ((mKP < MKaon + MPion) || (mKP > MBd - MPsi_nS) || (mPsiP < MPsi_nS + MPion) || (mPsiP > MBd - MKaon))
+    return 0.;
+    
+  // cos(theta(K*)) from masses
+  Double_t mKP2 = mKP*mKP;
+  Double_t mPsiP2 = mPsiP*mPsiP;
+  Double_t MPsi_nS2 = MPsi_nS*MPsi_nS;
+  Double_t _cKs = costhetaHel_host(MBd2, mKP2, MKaon2, MPion2 , MPsi_nS2, mPsiP2);
+  Double_t _cZc = costhetaHel_host(MBd2, mPsiP2, MPsi_nS2, MPion2, MKaon2, mKP2);
+  //cout <<"_cKs = " <<_cKs <<" for mKP2 = " <<mKP2 <<" and mPsiP2 = " <<mPsiP2 <<endl;
+    
+  if (fabs(_cKs) > 1 && _cZc > 1) {

```

Comment:
why
`fabs(_cKs) > 1 && _cZc > 1`
instead of
`fabs(_cKs) > 1 || fabs(_cZc) > 1`
?